### ISSUE TITLE
Please Relax global referencing rules for versioning

### ISSUE THREAD

**Issue by TehWardy** [dotnet/core#3760]
Hi guys,

I'm trying to use a lot of different parts of the net 2.2 core framework and aspnet and i'm finding constant conflicts due to referential constraints that you guys are imposing.

Can you internally pass around a rule that assemblies should require a minimum or something instead of requiring a specific version as the net result is 5 years or more of waiting for everything to sync up. 

For example ...
https://github.com/Azure/azure-functions-host/issues/4049#issuecomment-549612362
... and this sort of thing happens all over the place.

**Comment by leecow** [comment]
@TehWardy, can you describe your project and how you are encountering the issues? Adding a few folks to the discussion who may be able to help or re-direct as needed. 

@nguerrera @dsplaisted

**Comment by TehWardy** [comment]
So lets say i'm building a real world enterprise solution.
I begin by creating a few shared libraries that I compile against .Net standard, utility classes and what not. Then I layer on top of those a collection of complex applications with key .net / core dependencies.

So at this point I've already put together a solution that uses stuff like ...
 - EF Core.
 - Some Azure libs.
 - AspNet Core.
 - OData for that complex relational API layer.
 - SignalR for that real time stuff.
 - Specialized parts like the Roslyn API for complex tools.
 - Maybe a few third party tools. 

So having done all this and more I start to see issues putting all the pieces together.

The Azure functions package related to the ticket mentioned above will only accept Newtonsoft.Json package v11 with .Net core 2 usage and v9 or 10 i think it was with .Net usage. 
so if in my shared library that I began with happens to depend on v12 i now have to degrade that dependency or force the Azure function to use v12 to bring it up to date.

But then I find I might have some other part of .Net like the SignalR framework that decides only the latest version will do if provided otherwise it's preference is to use the new System.Text.Json API in that case i'm forcing a potential downgrade.

Which depending on the action I take with that might not then agree with what OData stuff I pulled in for the same reason,

Idea / suggestion here (simplest way I can see to improve this situation) ...

If the references in the framework as a rule usually only specified a runtime version or minimum version for dependencies then a lot of this type of issue goes away and the framework as a whole should become a bit more flexible about what it can be paired with.


Am I making sense here?

**Comment by dsplaisted** [comment]
In general the system works the way you suggest it should.  Versions don't need to match exactly.  The version loaded needs to be the same as or higher than the version a library was compiled against.

As I understand it, the Azure Functions issue is a more specific issue that has to do with plugins to a system.  The problem is that Azure Functions runs an ASP.NET Core app that then loads your code as a plugin.  By the time it loads your code, it's already loaded it's version of Newtonsoft.Json.  So then there's a versioning mismatch because the version of Newtonsoft.Json that is loaded is lower than the one required by your code (or its dependencies).

We've done some work to improve plugin support in .NET Core 3.0, so it might be possible now to have your code load a different version of Newtonsoft.Json than Azure Functions uses.  @jeffschwMSFT may know if this is the case.

It sounds like Azure Functions in this case may try to remove their dependency on Newtonsoft.Json instead.

**Comment by jeffschwMSFT** [comment]
> We've done some work to improve plugin support in .NET Core 3.0, so it might be possible now to have your code load a different version of Newtonsoft.Json than Azure Functions uses

We have a set of samples for this subset portion of the problem described above: https://github.com/dotnet/samples/tree/master/core/extensions/AppWithPlugin.  If you have any questions, please let me know.

**Comment by TehWardy** [comment]
Oh ok, thanks for the feedback.
I could just be unlucky and seemingly finding these edge cases because I use so much of the framework across my solution.

Some of my issues might be down to VS itself not playing ball too.

Am i right in understanding that .Net core 2.x and 3.x are not compatible as I can't seem to update to the latest EF Core stuff without it forcing me to move to .Net Core 3.0 which I'm a bit worried might be a little too close to the cutting edge for production code.

**Comment by TehWardy** [comment]
Ok here's another example that i'm suffering right now ...
https://github.com/Azure/azure-functions-host/issues/5230

... it's more dll hell and it's extremely frustrating.

The Azure function in question also uses the Roslyn API which seems to behave differently in the Function Host in the cloud to what it does on my local machine.

On top of that I attempted a full stack upgrade to .Net Core 3.0 and there seems to be no way to host SignalR with OData along side each other in an application as one requires the new endpoint routing stuff and the other just out right fails if it's enabled.

I'm now in a strange place where it seems like the signalR issue i'm having might be solved in .Net Core 3.0 (but i don't know) but even if it is I can't upgrade as i'm waiting on further improvements in OData before it'll all fit together.

Can I get some guidance on this please?

**Comment by TehWardy** [comment]
Ok I think I might be looking in the wrong place.
I've gone through a local server and noticed that during deploys IIS isn't stopping the kestrel process it's proxying and thus files are getting locked resulting a ton of un-predictable behaviour and god knows how many different issues.

On top of that I noticed that if I manually go to my projects bin folders after performing a clean solution in VS some files haven't been cleaned.
I don't know how to report VS issues but I would suggest that they completely wipe every projects bin folder during a clean since not doing so can often leave "other build configurations results" in the bin folder which i seem to be able to intermittently get it to pick up and use resulting in it thinking i'm trying to "debug a release deploy" or something.

I had to double check my release configurations to be sure that this was actually going on.

The net result is that it seems everything about putting a build together is what causes my issues, this got further confirmed when I finally fixed the publish profile on an azure function app and I deployed that and got the same exception I was seeing before even though locally I had resolved it.
I fixed the issue by completely removing the azure function instance through the cloud portal, deleting the publish profile then setting up a new one to deploy the function app to a resource with the same name again.

In short ... I can't be sure a "clean" deploy is what I think it is which complicates a lot of what i'm seeing.

Thinking about all these issues in combination I think there's a good chance that the framework is in fact behaving as described above and as designed, my issues seem to be largely around the cleanliness of builds. 

I have just one issue left to resolve which is around using the roslyn api's from inside an azure function but i have raised that separately.
I will leave this ticket open so you guys can decide what if anything you want to do with my findings / possible come back to me on it. 

Sorry about the confusion.

**Comment by carlossanlop** [comment]
Thanks for the details, @TehWardy . For now, I'll close it as it seems to be resolved.
If anyone else on this thread would like to provide additional comments, feel free to reopen to continue the conversation.
